[QNN] Requantize operator

anijain2305 opened the pull request:
Requantize converts one quantized tensor representation to another quantized
representation. The PR has following implementation features

Relevant Issue - #2351

- Requantize operator defined in qnn namespace - relay.op.qnn.requantize
- Lowering of the requantize to exisiting Relay operators
- Integer fixed point implementation of requantize
    - Two rounding modes - FE_UPWARDS (round towards infinity) and
    FE_AWAY_FROM_ZERO (std::round behavior)
- Floating point implementation as well, that can act as reference or can be
used for devices when FP32 computation is not expensive.
- Unit test cases

Credit to TFLite and GemmLowp to provide reference implementations.

Thanks for contributing to TVM!   Please refer to guideline https://docs.tvm.ai/contribute/ for useful information and tips. After the pull request is submitted, please request code reviews from [Reviewers](https://github.com/dmlc/tvm/blob/master/CONTRIBUTORS.md#reviewers).

tqchen commented:
Let us distinguish the qnn from the quantize namespace, given that the primary lowering phase supports qnn which is different from the current quantize namespace(automatic quantization in relay)

anijain2305 commented:
@tqchen Thanks for the quick comment :)

Can you please elaborate a little bit? Do you mean to change this file - https://github.com/dmlc/tvm/pull/3531/files#diff-dcae58edc1986609a54e3a202ee49b3b

anijain2305 commented on python/tvm/relay/quantize/rewrite.py:
```diff
@@ -0,0 +1,37 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
@tqchen Does it make sense to move this to newly created directory - `python/tvm/relay/op/qnn`?

This will ensure the `quantize` python namespace is not polluted.

anijain2305 commented on src/relay/pass/quantize_rewrite.cc:
```diff
@@ -0,0 +1,262 @@
+/*
```
@tqchen Should l call this file qnn_rewrite.cc to have a clear marking between `quantize` (Quantization in Relay) and qnn (creating new quantized ops) work?

anijain2305 commented:
@FrozenGene Please have a look.

anijain2305 commented on src/relay/pass/quantize_rewrite.cc:
```diff
@@ -0,0 +1,262 @@
+}
+
+RELAY_REGISTER_OP("qnn.requantize")
+.set_attr<FForwardRewrite>("FQuantizeForwardRewrite", RequantizeForwardRewrite);
```
@tqchen Does it make sense to replace FQuantizeForwardRewrite with FQnnForwardRewrite (for differentiating qnn and quantize)?

tqchen commented:
Given that qnn is a dialect. We could consider giving it a special name place and folder to live. e.g.

- src/relay/qnn
- src/relay/op/qnn
- tvm.relay.qnn

We can debate on whether it should be relay/qnn/op or relay/op/qnn  cc @jroesch  Perhaps let us open an RFC for dialect convention.

I have not yet taken a deep look into the PR, but here are a few high-level comments

- Header files, for utilities etc, we want to move most of them into src/ folder instead of include
   - The idea is that everything in the include is public API(which external project can depend on), everything else should be in src/
- There are a few style issues, we mainly use CamelCase for our APIs(as per Google C style).

anijain2305 commented:
I see. Makes sense. I don't have a strong preference for any of the namespaces that you suggested. My favorite is src/relay/qnn. In this case, we are only dealing with ops. But, I can believe that some other dialect might want to more than just ops. But, I am ok with any of the above options.
Please let me know. I can refactor to see how it looks like.

Thanks for the high-level comments. I will fix the utils and put them in the src/relay/qnn directory. Will look for CamelCase. (I was hoping that lint would do that for me, but I guess not).

anijain2305 commented:
@tqchen I have updated the code. The C++ files are in src/relay/qnn directory. It has its own sub-directories like include, op, pass to serve different purposes. Let me know how it looks.

FrozenGene commented on python/tvm/relay/op/qnn/qnn.py:
```diff
@@ -0,0 +1,74 @@
+from . import _make
+
+def requantize(input_data, input_zero_point, input_scale, output_zero_point,
+               output_scale, out_dtype="int32", use_int_compute=False,
```
The following relay attr's default value is `true`. Please unify the default value (We should set it be `True`).

anijain2305 replied:
Good catch. Thanks :)

FrozenGene commented on python/tvm/relay/op/qnn/qnn.py:
```diff
@@ -0,0 +1,74 @@
+
+def requantize(input_data, input_zero_point, input_scale, output_zero_point,
+               output_scale, out_dtype="int32", use_int_compute=False,
+               rounding_mode="FE_UPWARD"):
```
TFLite requires "round_awar_from_zero". Could you explain why we set it "FE_UPWARD" be the default value? Because of MXNet's requirement?

anijain2305 replied:
It can be "round_away_from_zero". These two rounding modes provide two points in the performance accuracy tradeoff curve. FE_UPWARD has fewer operators compared to round_way_from_zero. But, I think you are right, we should set "round_away_from_zero" by default. If somebody wants better performance, they can set it to FE_UPWARD using a network-level config.

FrozenGene commented on src/relay/qnn/pass/quantize_rewrite.cc:
```diff
@@ -0,0 +1,262 @@
+  // input_tensor, the result in int64 where the decimal point is sitting
+  // between bits 31 and 30 (from the right, rightmost bit is bit 0).
+  Expr scalar = MakeConstantScalar(up_idtype, fixed_point_multiplier);
+  auto multiplied_t = Multiply(tensor, scalar);
```
Where is `overflow` logic? If it is `overflow`, we return int32_max.

anijain2305 replied:
Will look into this. I remember seeing an overflow check. Thanks for pointing out.

anijain2305 replied:
@FrozenGene Thought little more about it. I think we don't need this check. Please let me know if this makes sense. 

The relevant portion from TFLite is
~~~
  bool overflow = a == b && a == std::numeric_limits<std::int32_t>::min();
~~~

Here a is input tensor and b is the fixed_point_multiplier. The fixed_point_multiplier is calculated using [frexp](https://en.cppreference.com/w/cpp/numeric/math/frexp).

Now, frexp gives negative fixed point multiplier only when the input floating point number is negative. In our case, the number is `input_scale/output_scale`. I think the scales are always positive, so we should never have negative number. (Let me know if I am wrong)

FrozenGene replied:
Right.

FrozenGene commented on src/relay/qnn/pass/quantize_rewrite.cc:
```diff
@@ -0,0 +1,262 @@
+ * The whole computation this can be broken down into following steps
+ * 1) Calculate the integer multiplier and integer shift.
+ * 2) Subtract the input integer point.
+ * 2) Multiply the integer fixed point multiplier with quantized tensor.
```
`2)` and `2)`

anijain2305 replied:
Good catch, will fix.

zhiics commented on src/relay/qnn/op/requantize.cc:
```diff
@@ -0,0 +1,91 @@
+ */
+
+/*!
+ *  Copyright (c) 2018 by Contributors
```
2019, or we should probably just remove it as we already have the license above.

Apply to other files as well.

zhiics commented on src/relay/qnn/pass/quantize_rewrite.cc:
```diff
@@ -0,0 +1,262 @@
+
+  // 2) Subtract the input_zero_point
+  auto tensor = input_tensor;
+  tensor = Cast(tensor, up_idtype);
```
`auto tensor = Cast(input_tensor, up_idtype);` ?

zhiics commented on src/relay/qnn/pass/quantize_rewrite.cc:
```diff
@@ -0,0 +1,262 @@
+    auto input_zp = MakeConstantScalar(up_idtype, param->input_zero_point);
+    tensor = Subtract(tensor, input_zp);
+  }
+
```
Remove blank lines

zhiics commented on src/relay/qnn/pass/quantize_rewrite.cc:
```diff
@@ -0,0 +1,262 @@
+  Expr scalar = MakeConstantScalar(up_idtype, fixed_point_multiplier);
+  auto multiplied_t = Multiply(tensor, scalar);
+
+
```
remove one blank line

zhiics commented on src/relay/qnn/pass/quantize_rewrite.cc:
```diff
@@ -0,0 +1,262 @@
+
+  // 4) Find the rounding scalar. This depends on where the final decimal point
+  // sits. As we will be right shifting the multiplied_t, we need to first
+  // calculate the totol_right_shift.
```
`total_right_shift`

zhiics commented on tests/python/unittest/test_quantized_ops.py:
```diff
@@ -0,0 +1,256 @@
+
+
+
+        if __name__ == "__main__":
```
move xx_test under `main` below directly? `run_test` looks not necessary

yzhliu replied:
how about separate them to individual `test_` functions instead of packing together to one giant `test_requantize`?

FrozenGene commented on tests/python/unittest/test_quantized_ops.py:
```diff
@@ -0,0 +1,256 @@
+        return func
+
+
+    def run_tests():
```
Add one `uint8` test. Like TFLite, whose out dtype is `uint8`. Let us cover it.

anijain2305 replied:
Good point. Will add it by tonight.

tqchen commented on src/relay/qnn/util.h:
```diff
@@ -0,0 +1,139 @@
+namespace tvm {
+namespace relay {
+
+inline bool IsInt8(const DataType& dtype) {
```
This is certainly a too small function to introduce another level of indirection. Directly remove it and keep things in the original condition

tqchen commented on src/relay/qnn/util.h:
```diff
@@ -0,0 +1,139 @@
+ */
+
+/*!
+ * \file tvm/relay/quantize_util.h
```
wrong file name

tqchen commented on src/relay/qnn/util.h:
```diff
@@ -0,0 +1,139 @@
+      || IsInt16(dtype) || IsUint16(dtype);
+}
+
+enum class QuantizeOpType : uint8_t {
```
unless there is a need(data compactness), you don't want to use uint8_t as storage type

tqchen commented on src/relay/qnn/util.h:
```diff
@@ -0,0 +1,139 @@
+  return dtype == Float(32);
+}
+
+inline bool IsQuantizedType(const DataType& dtype) {
```
IsQNNDataType

tqchen commented on src/relay/qnn/pass/quantize_rewrite.cc:
```diff
@@ -0,0 +1,262 @@
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
```
rename to qnn_lower.cc

tqchen commented on python/tvm/relay/qnn/op/qnn.py:
```diff
@@ -0,0 +1,74 @@
+from __future__ import absolute_import as _abs
+from . import _make
+
+def requantize(input_data, input_zero_point, input_scale, output_zero_point,
```
please debate on API naming, especially refer to possible existing API names and choices of options

tqchen commented:
I made a few quick reviews. @FrozenGene @zhiics @antinucleon @hlu1 please help check again if you have time. 

Let us specifically spend some time to discuss the API, in particular, what are the choices of parameter names, order and the options, do they make sense or not. Is there existing APIs that we can be consistent with etc.

yzhliu commented on include/tvm/relay/qnn/attrs.h:
```diff
@@ -0,0 +1,70 @@
+        .describe("Defines the rounding direction when the value is midway between"
+                  "two representable values. There are two supported modes - FE_UPWARD"
+                  "or FE_AWAY_FROM_ZERO. More context can be found at"
+                  "https://www.gnu.org/software/libc/manual/html_node/Rounding.html");
```
suggest to explain the meaning of the two modes here as well, so to make it self-explained.

anijain2305 replied:
Thanks, added the description.

yzhliu commented on python/tvm/relay/qnn/op/qnn.py:
```diff
@@ -0,0 +1,74 @@
+
+    Parameters
+    ----------
+    quantized_data : tvm.relay.Expr
```
doc does not match the arguments in the function

yzhliu commented on src/relay/qnn/util.h:
```diff
@@ -0,0 +1,139 @@
+  } else if (IsUint32(dtype)) {
+    return std::numeric_limits<uint32_t>::min();
+  }
+  LOG(FATAL) << "Type not supported\n";
```
also << dtype to make debug more easier. also I think we can remove "\n"

yzhliu commented on src/relay/qnn/pass/quantize_rewrite.cc:
```diff
@@ -0,0 +1,262 @@
+RELAY_REGISTER_OP("qnn.requantize")
+.set_attr<FForwardRewrite>("FQuantizeForwardRewrite", RequantizeForwardRewrite);
+
+TVM_REGISTER_API("relay._qnn.rewrite")
```
should this be ported to pass manager? @zhiics

zhiics replied:
Talked to @anijain2305 offline, it sounds this pass is more or less a dialect that is only performed in the frontend parsers. I think we probably don't want to bring it into the pass manager because 1) it is quite frontend specific, 2) it only does `Expr` -> `Expr` transformation.

We could probably use better naming instead of `rewrite` here and `ir_pass` in Python. Since we have `pass` under the dialect fold, should we have `transform` as well? 

@tqchen thoughts?

tqchen replied:
Yes, I think we need to use transform

yzhliu commented on src/relay/qnn/pass/quantize_rewrite.cc:
```diff
@@ -0,0 +1,262 @@
+    const Array<Expr>& new_args, const NodeRef& ctx) {
+  CHECK_EQ(new_args.size(), 1);
+  Expr quantized_data = new_args[0];
+  const auto* param = ref_call->attrs.as<RequantizeAttrs>();
```
if param == nullptr ?

yzhliu commented on src/relay/qnn/util.h:
```diff
@@ -0,0 +1,139 @@
+      return IsFloat32(in_dtype) || IsQuantizedType(in_dtype);
+    case QuantizeOpType ::Dequantize:
+      return IsQuantizedType(in_dtype);
+    case QuantizeOpType ::Requantize:
```
remove space before `::`

shoubhik commented:
> I made a few quick reviews. @FrozenGene @zhiics @antinucleon @hlu1 please help check again if you have time.
> 
> Let us specifically spend some time to discuss the API, in particular, what are the choices of parameter names, order and the options, do they make sense or not. Is there existing APIs that we can be consistent with etc.

Can you also look at #3512 from API name perspective as well.

anijain2305 commented:
Let me write down my initial pass over the API design. Please help me refine it further.

## Order of parameters
~~~
// All input relay exprs
input1_relay_expr,
input2_relay_expr,
...
...

// Properties for the input relay expr alphabetically (for example zero_point)
input1_prop1,
input1_prop2,
...
input2_prop1,
input2_prop2,

// Properties for the output relay expr alphabetically (for example zero_point)
output1_prop1,
output1_prop2,

// Finally, the properties of the ops alphabetically (like use_bias, use_int_compute)
op_prop1
op_prop2
~~~



## Parameter names

I could not find a specific pattern in the names. However, I think these are some common rules
* parameter name should be concise
* If there are multiple choices for the same param, most widely used one can be chosen to retain readability.
* for bool params, a good name can describe the action (E.g `count_include_pad`, `clip`). There are few examples in the codebase that do not follow this rule.

# Current proposal
Adhering to above rules, the API looks like this
~~~
def requantize(data,
               # Input expr properties
               input_scale,
               input_zero_point,

               # Output expr properties
               output_scale,
               output_zero_point,

               # Op properties
               use_int_domain, # compute was somewhat ambiguous
               rounding, # _mode seemed unnecessary
               out_dtype)
~~~

anijain2305 commented:
Thanks for all the comments. I fixed them. Please review once again.
Currently changed the API as per above proposal.

tqchen commented:
Can we also list the related APIs from existing frameworks e.g. tflite so we can get a good reference. Right now I can understand most of the parameters, except use_int_domain was a bit unclear to me

anijain2305 commented:
For TF, the [requantize](https://www.tensorflow.org/api_docs/cc/class/tensorflow/ops/requantize) operator looks like this

~~~
Arguments:

scope: A Scope object
input_min: The float value that the minimum quantized input value represents.
input_max: The float value that the maximum quantized input value represents.
requested_output_min: The float value that the minimum quantized output value represents.
requested_output_max: The float value that the maximum quantized output value represents.
out_type: The type of the output. Should be a lower bit depth than Tinput.
~~~

I think all the above arguments are represented in current API proposal (min/max are represented as scale/zero_point).
----

`use_int_domain` gives an ability to run the computation in FP32 domain. The integer domain has more instructions due to correct rounding, compared to FP32 computation. So, there can be a tradeoff for different HW targets.

However, if you think we will never use FP32 computation for requantize, that is also a reasonable answer.

FrozenGene commented:
Maybe we don't need `use_int_domain`. I can not imagine one HW we will set this be false.

anijain2305 commented:
Thanks @FrozenGene and @tqchen. Spent some more time thinking about it. Realized that it might not be necessary. Removed `use_int_domain`.

Please review again.

anijain2305 commented:
@tqchen Please let me know if you got a chance to look at the changes.

By the way, phisiart was added for review automatically when I was rebasing. Maybe, I made a mistake in rebasing. I don't know how to remove the reviewer.

tqchen commented on include/tvm/relay/qnn/attrs.h:
```diff
@@ -0,0 +1,75 @@
+        .describe("The scale of the input tensor.");
+    TVM_ATTR_FIELD(output_scale)
+        .describe("The scale of the output tensor.");
+    TVM_ATTR_FIELD(use_int_domain).set_default(true)
```
Please update the the attrs to remove use_int_domain.

tqchen commented on include/tvm/relay/qnn/attrs.h:
```diff
@@ -0,0 +1,75 @@
+  DataType out_dtype;
+
+  TVM_DECLARE_ATTRS(RequantizeAttrs, "relay.attrs.RequantizeAttrs") {
+    TVM_ATTR_FIELD(input_zero_point)
```
keep the argument order, and fields in the same order of the parameters

tqchen commented on include/tvm/relay/qnn/attrs.h:
```diff
@@ -0,0 +1,75 @@
+    TVM_ATTR_FIELD(out_dtype)
+        .set_default(NullValue<DataType>())
+        .describe("Output data type, set to explicit type under mixed precision setting");
+    TVM_ATTR_FIELD(rounding).set_default("FE_AWAY_FROM_ZERO")
```
The FE prefix is confusion, can we remove it?

tqchen commented on python/tvm/relay/qnn/ir_pass.py:
```diff
@@ -0,0 +1,37 @@
+
+from . import _qnn
+
+def qnn_lower(expr):
```
We are moving towards Module->Module pass, you can refer to @zhiics 's PR for more details

tqchen commented on src/relay/qnn/util.h:
```diff
@@ -0,0 +1,111 @@
+
+
+inline const int32_t GetQmax(const DataType& dtype) {
+  if (dtype == Int(8)) {
```
You can do it with bit arithmetics, without having to do switch. Note that DataType::max() also gives you this value in Expr.

tqchen commented:
Please also add a API docs as well as core ops. https://docs.tvm.ai/langref/relay_op.html we might need a section for qnn dialect.

anijain2305 commented:
@tqchen Thanks for the review. Incorporated your comments.

FrozenGene commented on python/tvm/relay/qnn/op/qnn.py:
```diff
@@ -0,0 +1,75 @@
+               output_scale,
+               output_zero_point,
+               rounding="AWAY_FROM_ZERO",
+               out_dtype="int32"):
```
`int32` maybe is not one good default value. In quantization computation, I can not imagine what condition will use `int32`. `uint8` (used in TFLite) or some quantization algorithm / framework's `int8` type maybe better.

anijain2305 replied:
Changed to int8.

FrozenGene commented on tests/python/unittest/test_qnn_ops.py:
```diff
@@ -0,0 +1,246 @@
+    return entry if isinstance(expr, relay.Function) else entry.body
+
+
+def test_requantize():
```
lack of `uint8` in some test.

anijain2305 replied:
Added a test in the latest commit.

shoubhik commented on src/relay/qnn/op/requantize.cc:
```diff
@@ -0,0 +1,94 @@
+
+}  // namespace qnn
+}  // namespace relay
+}  // namespace tvm
```
all cc files also need a blank line at the end

shoubhik commented on src/relay/qnn/pass/qnn_lower.cc:
```diff
@@ -0,0 +1,245 @@
+ *
+ * Use namespace to reduce potential naming conflict.
+ */
+namespace qnn_lower {
```
do we really need this namespace. should tvm/relay/qnn be not sufficient?

anijain2305 replied:
The Relay passes as of now have namespace for each pass. Ensuring that the code is similar for readability.

shoubhik commented on src/relay/qnn/util.h:
```diff
@@ -0,0 +1,108 @@
+namespace relay {
+namespace qnn {
+
+inline bool IsQNNDataType(const DataType& dtype) {
```
why not name it IsQuantizedDataType? `IsQNNDataType` is neither standard nor very easy to understand.

anijain2305 replied:
QNNDataType is mostly for differentiating the automatic quantization and QNN efforts.

anijain2305 commented on python/tvm/relay/qnn/__init__.py:
```diff
@@ -0,0 +1,21 @@
+"""Neural network related operators."""
+from __future__ import absolute_import as _abs
+from . import op
+from . import transform
```
The last line ends with a newline character. So, its fine.

anijain2305 commented on python/tvm/relay/qnn/_transform.py:
```diff
@@ -0,0 +1,22 @@
+from __future__ import absolute_import
+from tvm._ffi.function import _init_api
+
+_init_api("relay.qnn._transform", __name__)
```
The last line ends with a newline character. So, its fine.

anijain2305 commented on python/tvm/relay/qnn/op/__init__.py:
```diff
@@ -0,0 +1,20 @@
+# pylint: disable=wildcard-import
+"""Neural network related operators."""
+from __future__ import absolute_import as _abs
+from .qnn import *
```
The last line ends with a newline character. So, its fine.

anijain2305 commented on python/tvm/relay/qnn/transform.py:
```diff
@@ -0,0 +1,32 @@
+# specific language governing permissions and limitations
+# under the License.
+# pylint: disable=invalid-name
+"""QNN Dialect transformation passes."""
```
The last line ends with a newline character. So, its fine.

anijain2305 commented on python/tvm/relay/qnn/transform.py:
```diff
@@ -0,0 +1,32 @@
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+# pylint: disable=invalid-name
```
We need this globally in this particular file because the function names have to look similar to exisiting Relay passes. So, this should be fine here.

anijain2305 replied:
I will check other files for pylint

anijain2305 commented on python/tvm/relay/qnn/op/qnn.py:
```diff
@@ -0,0 +1,75 @@
+        The computed result.
+    """
+    assert rounding in ("UPWARD", "AWAY_FROM_ZERO"),\
+            "Unsupported rounding mode"
```
Moving to C++ and better error message

u99127 commented on include/tvm/relay/quantize_util.h:
```diff
@@ -0,0 +1,139 @@
+}
+
+
+
```
Unnecessary new lines.

u99127 commented on include/tvm/relay/quantize_util.h:
```diff
@@ -0,0 +1,139 @@
+  LOG(FATAL) << "Type not supported\n";
+  return -1;
+}
+
```
unnecessary new lines.

u99127 commented on include/tvm/relay/quantize_util.h:
```diff
@@ -0,0 +1,139 @@
+namespace tvm {
+namespace relay {
+
+inline bool is_Int8(const DataType& dtype) {
```
These should really be static inline to give a stronger hint for inlining. Further 'inline' is only a hint not a command, and given this is in a header file, it's safer to use static inline.

u99127 commented on src/relay/op/nn/requantize.cc:
```diff
@@ -0,0 +1,89 @@
+
+TVM_REGISTER_NODE_TYPE(RequantizeAttrs);
+
+
```
A comment here describing the function seems appropriate

u99127 commented on src/relay/pass/pattern_util.h:
```diff
@@ -373,6 +374,26 @@ inline Expr Copy(Expr data) {
 }
 
 
+inline Expr Where(const Expr& condition, const Expr& x, const Expr& y) {
```
s/inline/static inline

u99127 replied:
Replace inline with static inline.

anijain2305 replied:
Will skip this for now. There are other functions in this file, which need to be static inline. Will send out a separate patch for the whole file, if need be.

u99127 replied:
I would encourage you to do that with this PR itself because of the accidents missing this can cause :(. Alternatively I would suggest you follow up pretty quickly. 

You probably know this already but I'm documenting this for posterity.

TL;DR

This is one bit of c++ that I don't like TBH. I think you'll get away with it because there is just one name for that function in that namespace today. inline is a hint to the compiler and it's not mandatory for compilers to honour that request. Heuristics , eh :) , 

inline functions in c++ headers end up being implemented as weak functions in the translation unit. And then if you have multiple translation units containing such weak functions, one ends up with a case where the linker will pick an arbitrary definition depending on what got included where. Normally this isn't a problem if the same *definition* reaches everywhere, however all programmers are human (well, still :)) and are capable of choosing names that are similar to what's been chosen here. 

From my reading these functions are in a namespace that is common with other parts of the code and thus with the right set of conditions we could end up with a situation where someone is debugging a problem with this undefined behaviour from C++11 as that's the dialect we program in. 

regards
Ramana

anijain2305 replied:
Thanks, learnt something new today :)

Fair enough. I am adding `inline` for now in my added functions. We can create an Issue so that we ensure that we do it everywhere in the TVM codebase.

u99127 commented on src/relay/qnn/op/requantize.cc:
```diff
@@ -0,0 +1,94 @@
+namespace qnn {
+
+TVM_REGISTER_NODE_TYPE(RequantizeAttrs);
+
```
Missing comment

anijain2305 replied:
Added the comment below. Resolving this one.

u99127 commented on src/relay/qnn/pass/qnn_lower.cc:
```diff
@@ -0,0 +1,249 @@
+namespace qnn {
+/*!
+ * \brief namespace of qnn lower pass.
+ *
```
Suggestion : A brief description for the pass wouldn't go amiss here.

anijain2305 replied:
Thanks. There is a comment at the top of the file. I will explain it further there.

u99127 commented on src/relay/qnn/util.h:
```diff
@@ -0,0 +1,108 @@
+  Requantize
+};
+
+inline bool IsValidOpInputType(const QuantizeOpType& op_type,
```
s/inline/static inline.

u99127 commented on python/tvm/relay/qnn/op/qnn.py:
```diff
@@ -0,0 +1,73 @@
+    another quantized tensor representation. For the output tensor, we are
+    provided with output scale and zero point. The computation is as follows
+
+    Q_output = zp_output +  (scale_input)/(scale_ouptut) * (Q_input - zp_input)
```
type ouptut->output.

u99127 commented on src/relay/qnn/op/requantize.cc:
```diff
@@ -0,0 +1,94 @@
+  return true;
+}
+
+// Positional relay function to create quantized conv2d operator
```
Should we distinguish between 'quantized' and 'quantized conv2d' . I view 'quantized' as a generic concept vs a specific instantiation of 'quantized conv2d' .

u99127 commented on src/relay/qnn/pass/qnn_lower.cc:
```diff
@@ -0,0 +1,249 @@
+ *
+ */
+Expr RequantizeLower(const Expr& input_tensor,
+    const RequantizeAttrs* param, const DataType& idtype,
```
idtype == input Tensor data type ?

u99127 replied:
Shouldn't the inputs to this function get documented as per our coding standards  ?

anijain2305 replied:
Just added, resolving this one.

jackwish replied:
If *idtype The dtype of the input tensor*, we need this per TVM coding conversion?

u99127 commented on src/relay/qnn/pass/qnn_lower.cc:
```diff
@@ -0,0 +1,249 @@
+    tensor = Multiply(tensor, MakeConstantScalar(up_idtype, 1 << left_shift));
+  }
+  // Perform the multiplication in higher precision.
+  // If idtype is Int(32), the scalar is a fixed point value of int32 where the
```
The reason we are performing the multiplication in higher precision is to avoid overflow issues as the input tensor might be int32 and the scalar is a fixed point value of int32. Is that the thinking here ?

anijain2305 replied:
Yes

u99127 replied:
Let's be explicit in the comment then ?

anijain2305 replied:
Good point. Added at the end of the comment.

jackwish replied:
Wow, I am NOW aware that this patch is going to handle int32 data as requantize input. This will case problem in `GetFixedPointMultiplierShift()`.

anijain2305 commented:
@u99127 Thanks for going through the code in detail. I have added documentation for all C++ functions. Please let me know if it looks ok :)

u99127 commented on docs/langref/relay_op.rst:
```diff
@@ -196,6 +196,16 @@ This level support backpropagation of broadcast operators. It is temporary.
+
+   tvm.relay.qnn.op.requantize
+
+
```
I have more nits I'm sorry :) 
Extra newline.

anijain2305 replied:
I think this one seems fine. I just followed the style in this file. There are 2 new lines between each section.

u99127 commented on include/tvm/relay/qnn/attrs.h:
```diff
@@ -0,0 +1,71 @@
+                  "to -1). AWAY_FROM_ZERO is the standard rounding where the value"
+                  "is rounded away from zero at midpoints (for example, -1.5 rounds to"
+                  "-2). More context can be found at"
+                  "https://www.gnu.org/software/libc/manual/html_node/Rounding.html");
```
Minor nit and me being a grammar pedant :) 

s/At midpoint/At the midpoint. 

I would suggest being explicit with which rounding modes from the link this matches with from the glibc manual. Otherwise it will lead to a bit of head scratching which could be avoided.

u99127 commented on python/tvm/relay/qnn/__init__.py:
```diff
@@ -0,0 +1,21 @@
+# specific language governing permissions and limitations
+# under the License.
+# pylint: disable=wildcard-import
+"""Neural network related operators."""
```
More comment nits  ? Any thoughts of what is valid in this namespace ?

Quantized neural network related operators ?

u99127 commented on python/tvm/relay/qnn/op/qnn.py:
```diff
@@ -0,0 +1,73 @@
+# specific language governing permissions and limitations
+# under the License.
+#pylint: disable=invalid-name, too-many-lines
+"""Neural network operations."""
```
Quantized Neural network operators or is it operations ? 

Minor quibble.

u99127 commented on python/tvm/relay/qnn/op/qnn.py:
```diff
@@ -0,0 +1,73 @@
+        representable values.
+
+    out_dtype : str, optional
+        Specifies the output data type for mixed precision conv2d.
```
Remove mixed precision conv2d ?

u99127 commented on src/relay/qnn/pass/qnn_lower.cc:
```diff
@@ -0,0 +1,264 @@
+/*!
+ *  Copyright (c) 2019 by Contributors
+ * \file qnn_lower.cc
+ * \brief Lower qnn ops to a sequence of exisiting Relay ops.
```
s/exisiting/existing

u99127 commented on tests/python/unittest/test_qnn_ops.py:
```diff
@@ -0,0 +1,260 @@
+from tvm import relay
+from tvm.relay.testing import create_workload
+from tvm.contrib import graph_runtime
+
```
Are you planning in more of your qnn work to keep adding more unit tests to this one file or would you be adding separate files  ? If not, then please consider renaming this to test_qnn_requantize.py ? 

One of the things we will find as we keep adding tests is that it takes a long time to test one change as all the remaining tests need to be run. I don't yet grok the test framework but having multiple smaller tests might help us with test level parallelism considering the larger core counts these days. However that's an observation and not a requirement.

anijain2305 replied:
Having separate files seems like the right thing to do. Will rename the file.

u99127 commented on tests/python/unittest/test_qnn_ops.py:
```diff
@@ -0,0 +1,260 @@
+        mod = relay.qnn.transform.QnnLower()(mod)
+        return mod
+
+
```
Extra newline.

shoubhik commented on src/relay/qnn/pass/qnn_lower.cc:
```diff
@@ -0,0 +1,264 @@
+  DataType up_idtype = Int(2 * idtype_bits);
+
+  // 1) Calculating the integer multiplier and integer shift
+  std::pair<int, int> fixed_point_params =
```
in tflite there is a case 
```
if (double_multiplier == 0.) {
    *quantized_multiplier = 0;
    *shift = 0;
    return;
  }
```
i dont think we are handling this corner case?

anijain2305 replied:
Added the corner case. Thanks for pointing out.

shoubhik commented on src/relay/qnn/op/requantize.cc:
```diff
@@ -0,0 +1,102 @@
+  const Array<tvm::Expr> oshape = data->shape;
+  // assign output type
+  const RequantizeAttrs* param = attrs.as<RequantizeAttrs>();
+  reporter->Assign(types[1], TensorTypeNode::make(oshape, param->out_dtype));
```
we should also a dtype check for `out_dtype`. We need to make sure the `out_dtype` is integral type.

anijain2305 replied:
Good point! Adding a check.

shoubhik commented on python/tvm/relay/qnn/op/qnn.py:
```diff
@@ -0,0 +1,73 @@
+# specific language governing permissions and limitations
+# under the License.
+#pylint: disable=invalid-name, too-many-lines
+"""QNN dialect operators."""
```
Leave a blank like before this.

shoubhik commented on python/tvm/relay/qnn/op/qnn.py:
```diff
@@ -0,0 +1,73 @@
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#pylint: disable=invalid-name, too-many-lines
```
is this case i do not think there is a dange of "too-many-lines" as `max-module-lines=1500` seems to be configured.

u99127 commented on python/tvm/relay/qnn/__init__.py:
```diff
@@ -15,7 +15,7 @@
 # under the License.
 # pylint: disable=wildcard-import
-"""Neural network related operators."""
+"""QNN dialect operators and ir passes."""
```
Last one I hope :)

s/ir/IR

anijain2305 replied:
Done

anijain2305 commented:
> Hi Animesh,
> 
> Thanks a lot for fixing things up and your patient explanations !. I've suggested a few more changes that came to me as I read this again this evening , it looks much better but i've noticed a few other nits and a couple of places where the replacement of inline->static inline has been missed.
> 
> Many Thanks,
> Ramana

Ramana, I thank you for going through the details. Please do not hesitate to give more comments. This churn is very useful for producing good code :)

anijain2305 commented:
@u99127 and @shoubhik Thanks for all the comments again. I addressed them. You can click on "Files changed" tab at the top and resolve the comments that I have addressed in text, if they are good.

I would also encourage explicitly requesting comments and approve when you are happy with the code - https://docs.tvm.ai/contribute/code_review.html?highlight=explicitly#approve-and-request-changes-explicitly
This tells the the project committers that the requested comments have been addressed, instead of going through the comments.

jackwish commented on include/tvm/relay/qnn/attrs.h:
```diff
@@ -0,0 +1,73 @@
+        .describe("The scale of the output tensor.");
+    TVM_ATTR_FIELD(output_zero_point)
+        .describe("The zero point of the output tensor.");
+    TVM_ATTR_FIELD(rounding).set_default("AWAY_FROM_ZERO")
```
I suggest that we keep the GNU term `FE_TONEAREST` (maybe we can use a different prefix). I think by `AWAY_FROM_ZERO` you mean the midway points - where the *aways from zero* only take effects. But, it makes feel like it is against `FE_TOWARDZERO`, which means, when we have a -1.1, it will round to -2.0.

jackwish commented on src/relay/qnn/pass/qnn_lower.cc:
```diff
@@ -0,0 +1,271 @@
+    significand_int64 /= 2;
+    ++exponent;
+  }
+  CHECK_LE(significand_int64, std::numeric_limits<int>::max());
```
There is a problem here.

In TFLite, it checks the value range with `int32_t::max()` because it generates the `quantized_multiplier` with `int32_t` (I mean things like `auto q_fixed = static_cast<int64_t>(TfLiteRound(q * (1ll << 31)));`). So it is good.

However, if we are pushing to a more generic representation (though I don't know for what...), the value range check should consider `idtype_bits`, e.g. `2^(idtype_bits - 1)`. Also the casting, and the resulting multiplier and exponent. If, we have a `idtype_bits` of 64, the output could be meaningless.

Would you please help me to understand the point of a generic `idtype`?

jackwish commented on src/relay/qnn/pass/qnn_lower.cc:
```diff
@@ -0,0 +1,271 @@
+
+  // 1) Calculating the integer multiplier and integer shift
+  std::pair<int, int> fixed_point_params =
+      GetFixedPointMultiplierShift(double_multiplier, idtype);
```
I see this `idtype` could be of signed/unsigned integer of 8/16/32 bits. They could be safely handled by an `int32_t` representation, so are don't need the `idtype` parameter for `GetFixedPointMultiplierShift()`?

anijain2305 replied:
Thanks @jackwish for raising this point. Earlier I was trying to do minimal up_idtype. For example, `int8` input can just go to `int16`. But, your point that we can handle everything in `int32_t` is stronger.

I am changing the code to remove the dependence on in_dtype. We will always go to int32_t for any in_dtype. This will make code shorter and thus less prone to bugs. I already have tests with `int8`, so I can check the functionality as well.

Thanks for looking into such a detail :)

jackwish commented on src/relay/qnn/pass/qnn_lower.cc:
```diff
@@ -0,0 +1,271 @@
+  // 1) Calculating the integer multiplier and integer shift
+  std::pair<int, int> fixed_point_params =
+      GetFixedPointMultiplierShift(double_multiplier, idtype);
+  int fixed_point_multiplier = fixed_point_params.first;
```
do we really need to pack the pair and extract, what about the TFLite style?

anijain2305 replied:
Just seems like a better coding style. 
~~~
void QuantizeMultiplier(double double_multiplier, int32_t* quantized_multiplier,
                        int* shift) {
~~~

TFLite interface was this. Instead of passing them as pointers, I just created a pair.

I am cleaning up the code to use `std::make_pair` and `std::tie` to reduce the code bloat.

jackwish commented on src/relay/qnn/pass/qnn_lower.cc:
```diff
@@ -0,0 +1,271 @@
+  // 2) Subtract the input_zero_point
+  auto tensor = Cast(input_tensor, up_idtype);
+  if (param->input_zero_point != 0) {
+    auto input_zp = MakeConstantScalar(up_idtype, param->input_zero_point);
```
It would be a problem in practice, though the `up_idtype` could be 16 bit while the zero point are int32, I guess.

jackwish commented on src/relay/qnn/pass/qnn_lower.cc:
```diff
@@ -0,0 +1,271 @@
+  Expr round_scalar;
+  if (param->rounding == "UPWARD") {
+    auto pos_rounder = MakeConstantScalar(up_idtype, (1ll << (total_right_shift - 1)));
+    round_scalar = pos_rounder;
```
hmmm, `pos_rounder` is not needed?

jackwish commented on src/relay/qnn/pass/qnn_lower.cc:
```diff
@@ -0,0 +1,271 @@
+  tensor = multiplied_t;
+  Expr round_scalar;
+  if (param->rounding == "UPWARD") {
+    auto pos_rounder = MakeConstantScalar(up_idtype, (1ll << (total_right_shift - 1)));
```
Suggesting to name `1ll << (total_right_shift - 1)` and anything similar.

jackwish commented on src/relay/qnn/pass/qnn_lower.cc:
```diff
@@ -0,0 +1,271 @@
+  // clip_min and clip_max are within the dtype range of the input tensor to the
+  // clip operator. For example, if the input to clip operator is int8, but the
+  // out_dtype is uint8, we will get incorrect results, if we set max as 255.
+  auto q_min = std::max(GetQmin(param->out_dtype), GetQmin(idtype));
```
The `GetQmin` and `GetQmax` should use output data type?

jackwish commented on src/relay/qnn/pass/qnn_lower.cc:
```diff
@@ -0,0 +1,271 @@
+  CHECK(param != nullptr);
+
+  // Find output shape.
+  Array<IndexExpr> out_shape;
```
need not to declare so early?

anijain2305 commented on include/tvm/relay/qnn/attrs.h:
```diff
@@ -0,0 +1,73 @@
+        .describe("The scale of the output tensor.");
+    TVM_ATTR_FIELD(output_zero_point)
+        .describe("The zero point of the output tensor.");
+    TVM_ATTR_FIELD(rounding).set_default("AWAY_FROM_ZERO")
```
Good point.

anijain2305 commented:
> Basically good, with some minor arithmetic issues. Please ping me if I understand in wrong way.

Thanks @jackwish for the review. I updated the code with your suggestions. Another review will be helpful.

jackwish commented on src/relay/qnn/util.h:
```diff
@@ -0,0 +1,113 @@
+static inline bool IsQNNDataType(const DataType& dtype) {
+  return dtype == Int(8) || dtype == UInt(8)
+      || dtype == Int(16) || dtype == UInt(16)
+      || dtype == Int(32);
```
I suggest that we re-evaluate whether do we need 32 bit quantization support. Let's discuss in the RFC thread to loop more people.

anijain2305 replied:
Is there any particular reason we should not include Int(32) for requantize?

I think for requantize, the in_dtype can be `int32`. For out_dtype, it seems unlikely that it will be `int32`.

jackwish replied:
I think requantize is fine as the intermediate results are int32. My consideration is quantize and requantize. I will leave comments to the RFC #3591 later :)

jackwish replied:
see https://github.com/dmlc/tvm/issues/3591#issuecomment-515008567

jackwish commented on src/relay/qnn/pass/qnn_lower.cc:
```diff
@@ -0,0 +1,254 @@
+ *
+ *       The whole computation this can be broken down into following steps
+ *       1) Calculate the integer multiplier and integer shift.
+ *       2) Subtract the input integer point.
```
"integer zero point"?

jackwish commented on src/relay/qnn/pass/qnn_lower.cc:
```diff
@@ -0,0 +1,254 @@
+ *       The whole computation this can be broken down into following steps
+ *       1) Calculate the integer multiplier and integer shift.
+ *       2) Subtract the input integer point.
+ *       3) Multiply the integer fixed point multiplier with quantized tensor.
```
"integer" may have assumed "fixed point" already.

jackwish commented on src/relay/qnn/pass/qnn_lower.cc:
```diff
@@ -0,0 +1,254 @@
+    auto zero_t = Full(zero, out_shape, hp_dtype);
+    round_scalar = Where(GreaterEqual(tensor, zero_t), pos_rounder_t,
+            neg_rounder_t);
+  }
```
Do we need to handle illegal rounding policy here? Just curious, maybe it has already been handled.

anijain2305 replied:
We do check that at Line 220 below in the same file..

jackwish commented on src/relay/qnn/pass/qnn_lower.cc:
```diff
@@ -0,0 +1,254 @@
+  auto q_min = GetQmin(param->out_dtype);
+  auto q_max = GetQmax(param->out_dtype);
+  auto clipped_t = Clip(shifted_int64_t, q_min, q_max);
+  auto requantized_output = Cast(clipped_t, param->out_dtype);
```
maybe return directly?

jackwish commented on src/relay/qnn/util.h:
```diff
@@ -0,0 +1,99 @@
+namespace qnn {
+
+enum class QuantizeOpType {
+  Quantize,
```
Do we need to move Quantize/Dequantize all to another patch?

jackwish commented on src/relay/qnn/util.h:
```diff
@@ -0,0 +1,99 @@
+  Requantize
+};
+
+static inline bool IsValidOpInputType(const QuantizeOpType& op_type,
```
Maybe we can inline them into where to call it, it will be more easy to read and debug I guess.

u99127 replied:
With recent compilers and debuggers stepping in and out of inline functions has become quite better, so I'm less fussed about it.  

To mind  the abstraction of the checker function, has the advantage that changing this on a later date means we don't have to go change all the places which are needed in case we care about higher order quantization. Actually I feel that makes the code more readable.

anijain2305 replied:
Each Relay operator has a InferType function, whose responsibility is to check/set the shapes and dtypes. So, it seems like to me that there is an already a good abstraction. So, I decided to move the check to that Infer function.

anijain2305 commented:
@tqchen I think we have couple of approvals about the API design and overall PR. Can you please make a pass and let me know if you have more comments here?

u99127 commented on src/relay/qnn/util.h:
```diff
@@ -62,19 +52,15 @@ static inline bool IsValidOpInputType(const QuantizeOpType& op_type,
 
 static inline const int32_t GetQmin(const DataType& dtype) {
-  CHECK(IsQNNDataType(dtype))
+  CHECK_LE(dtype.bits(), 32)
```
The types are discrete, but the Check here is CHECK_LE which means that technically this allows 16 bit values ?

anijain2305 replied:
Yes this is equivalent to num_of_bits <= 32

anijain2305 commented:
@tqchen Ping if you missed the last one :)

tqchen commented:
Thanks @anijain2305 , sorry for the delay as I was busy this week defending my thesis. Await approval from @FrozenGene @u99127

anijain2305 commented:
@tqchen Congrats :) I am certain that you would have aced the defense. Please keep doing the awesome work :)

@FrozenGene Please let me know if it looks good.

FrozenGene commented:
> @tqchen Congrats :) I am certain that you would have aced the defense. Please keep doing the awesome work :)
> 
> @FrozenGene Please let me know if it looks good.

I will review this today afternoon (Beijing Time).

tqchen commented on docs/langref/relay_op.rst:
```diff
@@ -198,6 +198,16 @@ This level support backpropagation of broadcast operators. It is temporary.
    tvm.relay.contrib.adaptive_avg_pool2d
 
 
+**Level 11: QNN Dialect Operators**
```
Let us use 11 for all dialect ops

tqchen commented on python/tvm/relay/qnn/transform.py:
```diff
@@ -0,0 +1,33 @@
+
+from . import _transform
+
+def QnnLower():
```
To make terminology consistent with LLVM/MLIR let us call it Legalize.

tqchen replied:
https://wiki.aalto.fi/display/t1065450/LLVM+Legalization+and+Lowering

tqchen replied:
Possibly can be handled in https://github.com/dmlc/tvm/pull/3672

anijain2305 replied:
Not sure if Legalize is the right term here. Legalize seems to target-dependent pass. QNN ops would be lowered to a sequence of Relay ops. Those Relay ops might go through Legalize API if need be.

tqchen replied:
I am also not 100% sure. However, we could view the relay core ops as a target and legalization translate the qnn dialect into the core ops.

FrozenGene replied:
I think `Lower` maybe is one better name. `Legalize` in LLVM means when we have one `type` (such as `i64` in `x86_32`) or instruction (such as `ISD::OR` of `v4i32` in `x86` backend), we make these types or instruction can be supported in one specific backend. `Lower` means we make high-level concept lower into low level instruction / op combine, i.e. the high-level concept (`qnn` here ) is not supported on **all** target and we should lower it into sequence of low level op. Maybe some low level op is not supported by one specific backend, then we legalize it.

tqchen commented on include/tvm/relay/qnn/attrs.h:
```diff
@@ -0,0 +1,71 @@
+        .describe("The scale of the output tensor.");
+    TVM_ATTR_FIELD(output_zero_point)
+        .describe("The zero point of the output tensor.");
+    TVM_ATTR_FIELD(rounding).set_default("TONEAREST")
```
Please double check the choice of enums "TONEAREST", are there existing API choices that are similar? I do not have preference but would be great to survey the related APIs should it be "TO_NEAREST"? (The TF API convention seems to be TO_NEAREST)

anijain2305 replied:
I had similar question while coding this. I used TONEAREST because the supporting document - https://www.gnu.org/software/libc/manual/html_node/Rounding.html
had FE_TONEAREST. To be consistent with that document, I kept TONEAREST. But, I like TO_NEAREST better as it aligns with the rest of the codebase.

tqchen commented:
@anijain2305 seems that the only todo point was the naming for QNNLower, given that we could simply use the Legalize PR to implement it, do you mind remove it from this PR and we can merge the rest part in?

anijain2305 commented:
@tqchen Sorry for the delay, I was on a vacation. Ok, give me a couple of days and I will work on it.

tqchen commented:
@anijain2305 no problem, enjoy your holiday and take your time. just ping the thread when you think things are ready

anijain2305 commented:
@tqchen Moved to Legalize API. Please review if things look ok now.

FrozenGene commented on src/relay/qnn/op/requantize.cc:
```diff
@@ -0,0 +1,279 @@
+ *
+ * Q_output = zp_output +  (scale_input)/(scale_ouptut) * (Q_input - zp_input)
+ */
+Expr RequantizeLegalize(const Attrs& attrs, const Array<Expr>& new_args,
```
Hmm...Could we discuss this more about whether to use `Legalize` or `Lower`. As replied yesterday, IMO, `Lower` is better. Because this is target-independent.

anijain2305 replied:
Actually the code looks cleaner with Legalize. I understand your concern when you say that Legalize is a target-dependent pass. However, we can think of target here as Relay ops with TVM Compute, where the QNN ops do not have any compute. Its a little stretch but the API will be exactly same, so maybe it is ok to use the Legalize API.

anijain2305 commented:
@tqchen Can we try to get this merged if everything looks good? Will open up discussion/review for many other PRs.

tqchen commented:
Thanks @anijain2305 @u99127 @fronzenGene, @yzhliu @jackwish @zhiics @shoubhik
